crates/lib/src/qvm/mod.rs

```diff
@@ -32,73 +41,125 @@ impl QvmResultData {
-            trials: shots,
-            request_type: RequestType::Multishot,
+    // Create a clone of the program so MOVE statements can be prepended to it
+    let mut program = program.clone();
+
+    for (name, values) in params {
+        match program.memory_regions.get(name.as_ref()) {
+            Some(region) => {
+                if region.size.length != values.len() as u64 {
+                    return Err(Error::RegionSizeMismatch {
+                        name: name.clone(),
+                        declared: region.size.length,
+                        parameters: values.len(),
+                    });
+                }
+            }
+            None => {
+                return Err(Error::RegionNotFound { name: name.clone() });
+            }
+        }
+        for (index, value) in values.iter().enumerate() {
+            program.instructions.insert(
+                0,
+                Instruction::Move(Move {
+                    destination: ArithmeticOperand::MemoryReference(MemoryReference {
+                        name: name.to_string(),
+                        index: index as u64,
+                    }),
+                    source: ArithmeticOperand::LiteralReal(*value),
+                }),
+            );
         }
     }
-}
 
-#[derive(Serialize, Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
-#[serde(rename_all = "lowercase")]
-enum RequestType {
-    Multishot,
+    let request = api::MultishotRequest::new(
+        &program.to_string(true),
+        shots,
+        addresses,
+        measurement_noise,
+        gate_noise,
+        rng_seed,
+    );
+    api::run(&request, config)
+        .await
+        .map(|response| QvmResultData::from_memory_map(response.registers))
 }
 
-#[cfg(test)]
-mod describe_request {
-    use std::borrow::Cow;
-
-    use super::Request;
-
-    #[test]
-    fn it_includes_the_program() {
-        let program = "H 0";
-        let request = Request::new(program, 1, &[]);
-        assert_eq!(&request.quil_instructions, program);
-    }
-
-    #[test]
-    fn it_uses_kebab_case_for_json() {
-        let request = Request::new("H 0", 10, &[Cow::Borrowed("ro")]);
-        let json_string = serde_json::to_string(&request).expect("Could not serialize QVMRequest");
-        assert_eq!(
-            serde_json::from_str::<serde_json::Value>(&json_string).unwrap(),
-            serde_json::json!({"type": "multishot", "addresses": {"ro": true}, "trials": 10, "quil-instructions": "H 0"})
-        );
-    }
+/// All of the errors that can occur when running a Quil program on QVM.
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Error parsing Quil program: {0}")]
+    Parsing(#[from] ProgramError<Program>),
+    #[error("Shots must be a positive integer.")]
+    ShotsMustBePositive,
+    #[error("Declared region {name} has size {declared} but parameters have size {parameters}.")]
+    RegionSizeMismatch {
+        name: Box<str>,
```

Comment:
Why a `Box<str>`? Does that give us any benefit over a `String`?

Reply:
I dunno, I didn't write it, just moved it out of the API module. If I had to guess, it's because the only place this error gets initialized, name comes from a map using `Box<str>`. I can change it if you'd like?